[Merged by Bors] - feat(Tactic/WLOG): add `wlog!` a variant of `wlog` that calls `push_neg`

Hagb opened the pull request:
This commit adds `wlog!`, a variant of the `wlog h : P` tactic that also
calls `push_neg` at the generated hypothesis `h : ¬ p` in the side goal.

---

- [x] depends on: #31406

<!-- Your PR title will become the first line of the commit message.

In this box, the text above the `---` (if not empty) will be appended
to the commit message, and can be used to give additional context or
details. Please leave a blank newline before the `---`, otherwise GitHub
will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Hagb committed:
feat(Tactic/WLOG): add `wlog!` a variant of `wlog` that calls `push_neg`

JovanGerb reviewed (commented):
Thanks for making this PR!

Unfortunately this clashes with #31406, so I think it is better to wait for that one first.

Hagb committed:
Merge remote-tracking branch 'upstream/master' into wlog

Hagb committed:
apply new `push_neg`

Hagb committed:
add a test case of `+distrib`

JovanGerb commented on MathlibTest/wlog.lean:
```diff
@@ -170,3 +170,9 @@ example (x y : ℕ) : True := by
   · trivial
+
+example (x y : ℕ) : True := by
+  wlog! +distrib h: x = 0 ∧ y = 0
```
```suggestion
  wlog! +distrib h : x = 0 ∧ y = 0
```

Hagb commented on Mathlib/Tactic/WLOG.lean:
```diff
@@ -114,6 +116,24 @@ def _root_.Lean.MVarId.wlog (goal : MVarId) (h : Option Name) (P : Expr)
+  if push then
+    reductionGoal.withContext do
+      let negHygName := mkIdent <| ← reductionFVarIds.2.getUserName
+      evalTactic (← `(tactic| try push_neg at $negHygName:ident))
```
Thanks!

JovanGerb reviewed (commented):
Thanks!

maintainer delegate

Hagb committed:
syntactically match hypotheses except `push_neg`'ed ones with `<` or `≤`

There may be instance diamond on `<` or `≤` so they may not be
syntactically equal.

Hagb committed:
Update Mathlib/Tactic/WLOG.lean

Co-authored-by: Jovan Gerbscheid <56355248+JovanGerb@users.noreply.github.com>

jcommelin reviewed (commented):
Thanks :tada:

bors merge
